Only include apps that are available for the restoring user when generating reports

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Depends on https://github.com/dimagi/commcare-hq/pull/34546 (currently based off of that branch), and will ultimately replace https://github.com/dimagi/commcare-hq/pull/34532.

It is likely the case that this will become the accepted way of restoring reports since it isn't necessary to include apps that a user does not have access to. However for now this is behind a feature flag for ease of enabling/testing.

For now, because we cannot reliably determine if a restore was triggered by web apps or live preview, we have to check if the underlying web user has permission to view apps. If they do, we cannot safely limit the reports included in a restore since it could be a live preview context where web apps permissions do not apply.
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
RESTORE_ACCESSIBLE_REPORTS_ONLY

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
The change is entirely behind a feature flag which certainly limits the blast radius and makes it easy to turn off if needed.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Create toggle to restore accessible reports only

gherceg committed:
Only use accessible apps when restoring mobile reports

gherceg committed:
Fix lint errors and run isort

esoergel commented on corehq/apps/app_manager/fixtures/mobile_ucr.py:
```diff
@@ -112,6 +111,11 @@ def _get_apps(self, restore_state, restore_user):
 
         if app_aware_sync_app:
             apps = [app_aware_sync_app]
+        elif toggles.RESTORE_ACCESSIBLE_REPORTS_ONLY and restore_state.params.is_webapps:
```
This still needs a `and not request_user.can_edit_apps` or something, right?

gherceg replied:
Yeah sorry just pushed the changes we had talked about. Turns out we need to use the `view_apps` permission as the indicator since a user who can view apps would be able to use live preview.

gherceg committed:
Only limit apps if request user cannot edit apps

gherceg commented:
Looks like the permission we care about is "view_apps" which unfortunately reduces the impact of this optimization even further, but should still apply for the projects that are strict about app access.

gherceg committed:
Check if user can view apps

A user that can view apps could access restores via live preview, which
means we need to sync all apps since web apps access cannot be reliably
used in this context.

gherceg committed:
Check if the toggle is actually enabled

esoergel commented on corehq/apps/app_manager/fixtures/mobile_ucr.py:
```diff
@@ -112,6 +111,18 @@ def _get_apps(self, restore_state, restore_user):
+            and restore_state.params.is_webapps
+        ):
+            apps = []
+            if restore_user.request_user.can_view_apps():
```
What would you think of moving this conditional into the `elif` above?  Feels like that might more clearly portray that the toggle only applies to web apps users who can't view apps, and everyone else gets the `else` clause below (so it's clear it's not just a coincidence that lines 121 and 127 are the same).

gherceg replied:
Yeah I can get behind that. My rationale was to clearly show the ambiguous scenario where we can't confidently determine if the restore is from web apps or live preview, and fallback to the default behavior, but to your point it isn't clear that it falls back to the default behavior.

gherceg committed:
Move can_view_apps check up one level

gherceg committed:
Pass domain into can_view_apps check

Also, should be if user cannot view apps

esoergel commented on corehq/apps/app_manager/fixtures/mobile_ucr.py:
```diff
@@ -114,7 +114,8 @@ def _get_apps(self, restore_state, restore_user):
             and restore_state.params.is_webapps
-            and restore_user.can_view_apps()  # only way to be sure this a web apps restore, not live preview
+            # only way to reliably know that this is a web apps restore, not live preview
+            and not restore_user.can_view_apps(restore_user.domain)
```
oof yeah that one's a gotcha.  I wonder what it'd take to make `domain` a required arg here

gherceg committed:
Use the request user to determine permissions

millerdev commented on corehq/apps/app_manager/fixtures/mobile_ucr.py:
```diff
@@ -112,6 +111,15 @@ def _get_apps(self, restore_state, restore_user):
+        ):
+            for app in get_web_apps_available_to_user(restore_user.domain, restore_user._couch_user):
+                if not is_remote_app(app):
+                    apps.append(get_correct_app_class(app).wrap(app))
```
Where is `apps` initialized? I see it in the `if app_aware_sync_app:` branch above, but if that executes this branch will not, I think.

Maybe a test or two would be good for this function?

gherceg committed:
Initialize apps
